### PR TITLE
Merge 'cosign/pkg/providers' into sigstore/sigstore.

### PR DESCRIPTION
**Pull request by wlynch**
#### Summary

Merge 'cosign/pkg/providers' into sigstore/sigstore.

This upstream's cosign's provider packages into `sigstore/sigstore` so that other tools like `gitsign` can use them without needing to depend on cosign. These packages are generic enough that they seem like a good fit here to be shared across sigstore projects.

This was intentionally done as a merge so that commit history is preserved for both the sigstore repo and the commits that are coming from cosign.

That said, we could also do this as a rebase on top of sigstore (would rewrite all commit times for cosign commits) or a single squash commit (all changes attributed to 1 commit). Let me know if you have thoughts on this.

#### Ticket Link

Part of https://github.com/sigstore/gitsign/issues/62

#### Release Note

```release-note
cosign's OAuth provider packages are now available in sigstore/sigstore.
```

### CODE REVIEW

**Comment by wlynch** [comment]
cc @imjasonh

**Comment by cpanato** [comment]
@wlynch please sign the DCO :)

File: pkg/oauth/providers/doc.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/doc.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/all/all.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/all/all.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/filesystem/doc.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/filesystem/doc.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/filesystem/filesystem.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/filesystem/filesystem.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/github/doc.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/github/doc.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/github/github.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/github/github.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/google/doc.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/google/doc.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/google/google.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/google/google.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/interface.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/interface.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/spiffe/doc.go [hunk 10]
```diff
@@ -0,0 +1,17 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/spiffe/doc.go:2]:
```suggestion
// Copyright 2022 The Sigstore Authors.
```

File: pkg/oauth/providers/spiffe/spiffe_test.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2022 The Sigstore Authors.
```
Comment (cpanato) [inline pkg/oauth/providers/spiffe/spiffe_test.go:1]: nit: add an extra `//` in the beginning of the file

**Review by cpanato — COMMENTED** [review]
small boilerplate changes

**Comment by imjasonh** [comment]
This is great!

Longer term I'd like to think about splitting these out into separate repos (along with KMSes), and eliminating 'All` as an antipattern. Otherwise we run the risk of just making s/s the same huge dependency bomb that s/c is.

But this is a clear improvement over the cosign dependency bomb, so let's do it!

**Comment by wlynch** [comment]
> @wlynch please sign the DCO :)

I actually did sign the DCO 😅  - the errors are coming from the older commits, even though they also have DCOs attached. This is also why the copyright notices say 2021 - because they're being preserved from the old files in https://github.com/sigstore/cosign/tree/main/pkg/providers

Any ideas for how to resolve this? Or should I just give up and obliterate the history?

**Comment by cpanato** [comment]
You can rebase with the interactive mode and signoff each commit

**Comment by wlynch** [comment]
> You can rebase with the interactive mode and signoff each commit

Done! Do will still need to update the copyright years, or is it alright to leave as is?

**Comment by dlorenc** [comment]
One last merge issue in go.mod

**Comment by imjasonh** [comment]
This also starts addressing https://github.com/sigstore/cosign/issues/1867 🎉

**Comment by dlorenc** [comment]
Is this one ready to go?

**Comment by imjasonh** [comment]
I wonder if we should wait and just move these out to separate repos. That will result in a simpler history than having to go from e.g., sigstore/oidc-google -> sigstore/sigstore -> sigstore/cosign.

I don't feel strongly either way, and I don't know how soon we'll be able to get consensus about where these packages should live long-term, so we shouldn't block on deciding this.

**Comment by haydentherapper** [comment]
@wlynch Are we good to close this and punt til sigstore-go?

**Comment by haydentherapper** [comment]
Closing in favor of sigstore-go

**Pull request closed without merging**